MISRA-C evaluate boolean operands

ceolin opened the pull request:
This is a massive change, most problems were spotted by the coccinelle script in the first commit.

ceolin commented:
issue #9889

ceolin commented on lib/libc/minimal/source/stdio/fopen.c:
```diff
@@ -0,0 +1,54 @@
+/* stdout_console.c */
```
arghh, my mistake. I added this file by mistake, this was something that I was working and forget to stash.

ceolin commented on kernel/sched.c:
```diff
@@ -782,7 +782,7 @@ void _impl_k_yield(void)
 	__ASSERT(!_is_in_isr(), "");
 
-	if (!_is_idle(_current)) {
+	if (_is_idle(_current)) {
```
double check it. Thanks !

andrewboie replied:
I don't understand your response.

_is_idle() is declared in this same C file, and there are no changes to it.

ceolin replied:
sorry, I meant that I'd double check it

ceolin commented on kernel/sched.c:
```diff
@@ -180,7 +180,7 @@ static struct k_thread *next_up(void)
 		}
 
-		if (!should_preempt(th, _current_cpu->swap_ok)) {
+		if (should_preempt(th, _current_cpu->swap_ok)) {
```
ditto

andrewboie replied:
ditto

should_preempt returns an int by the way, it should be changed to a bool

andrewboie commented:
If patch 3 was only partially machine-generated, I'd suggest splitting the third patch into two patches: one purely the output of coccinelle doing its thing, and then another one with the manual tweaks.

andrewboie commented on lib/posix/pthread.c:
```diff
@@ -493,7 +493,7 @@ int pthread_attr_setdetachstate(pthread_attr_t *attr, int detachstate)
 int pthread_attr_getschedpolicy(const pthread_attr_t *attr, int *policy)
 {
-	if (!attr || !attr->initialized) {
+	if ((attr == NULL) || (attr->initialized == NULL)) {
```
I don't think attr->initialized is a pointer

andrewboie commented on subsys/cpp/cpp_virtual.c:
```diff
@@ -19,7 +19,7 @@
 void __cxa_pure_virtual(void)
 {
-	while (1) {
+	while (true) {
```
does C++ define "true"? Might need to pull in an include somwhere

andrewboie replied:
```
FAILED: ccache /opt/sdk/zephyr-sdk-0.9.5/sysroots/x86_64-pokysdk-linux/usr/bin/xtensa-zephyr-elf/xtensa-zephyr-elf-gcc -DBUILD_VERSION=zephyr-v1.13.0-2185-g9d859ae904 -DKERNEL -D_FORTIFY_SOURCE=2 -D__ZEPHYR__=1 -I../../../../../../kernel/include -I../../../../../../arch/xtensa/include -I../../../../../../soc/xtensa/sample_controller -I../../../../../../soc/xtensa/sample_controller/include -I../../../../../../soc/xtensa/include -I../../../../../../include -I../../../../../../include/drivers -Izephyr/include/generated -isystem /opt/sdk/zephyr-sdk-0.9.5/sysroots/x86_64-pokysdk-linux/usr/lib/xtensa-zephyr-elf/gcc/xtensa-zephyr-elf/6.2.0/include -isystem /opt/sdk/zephyr-sdk-0.9.5/sysroots/x86_64-pokysdk-linux/usr/lib/xtensa-zephyr-elf/gcc/xtensa-zephyr-elf/6.2.0/include-fixed -I../../../../../../lib/libc/minimal/include -I../../../../../../tests/include -I../../../../../../tests/ztest/include -I/opt/sdk/zephyr-sdk-0.9.5/sysroots/xtensa-zephyr-elf/usr/include/arch/include -I/opt/sdk/zephyr-sdk-0.9.5/sysroots/xtensa-zephyr-elf/usr/include -Os -g -Wall -Wformat -Wformat-security -Wno-format-zero-length -imacros /home/buildslave/src/github.com/zephyrproject-rtos/zephyr/out-2nd-pass/qemu_xtensa/tests/application_development/cpp/test/zephyr/include/generated/autoconf.h -ffreestanding -Wno-main --sysroot /opt/sdk/zephyr-sdk-0.9.5/sysroots/xtensa-zephyr-elf/usr -Werror -Wno-deprecated-declarations -fno-asynchronous-unwind-tables -fno-pie -fno-pic -fno-strict-overflow -Wno-pointer-sign -Wno-unused-but-set-variable -fno-reorder-functions -fno-defer-pop -Werror=implicit-int -Wpointer-arith -ffunction-sections -fdata-sections -mlongcalls -std=c99 -MD -MT zephyr/CMakeFiles/zephyr.dir/subsys/cpp/cpp_virtual.c.obj -MF zephyr/CMakeFiles/zephyr.dir/subsys/cpp/cpp_virtual.c.obj.d -o zephyr/CMakeFiles/zephyr.dir/subsys/cpp/cpp_virtual.c.obj   -c /home/buildslave/src/github.com/zephyrproject-rtos/zephyr/subsys/cpp/cpp_virtual.c
/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/subsys/cpp/cpp_virtual.c: In function __cxa_pure_virtual:
/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/subsys/cpp/cpp_virtual.c:22:9: error: true undeclared (first use in this function)
  while (true) {
         ^~~~
/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/subsys/cpp/cpp_virtual.c:22:9: note: each undeclared identifier is reported only once for each function it appears in
ninja: build stopped: subcommand failed.
```

Think the problem is that this file is being compiled by the C and not C++ compiler (note filename extension)

andyross replied:
FWIW, the C++ language defines "true" as a builtin token.  In C, the official standard (for backward compatibility reasons) defines "_True" as the builtin, and you're supposed to include <stdbool.h> to get the symbol that everyone actually uses.

pfalcon commented on arch/arc/core/cache.c:
```diff
@@ -58,7 +58,7 @@ static int dcache_available(void)
 static void dcache_dc_ctrl(u32_t dcache_en_mask)
 {
-	if (!dcache_available())
+	if (dcache_available() == 0)
```
+1

aescolar reviewed (commented):
I don't have a problem with the changes to the posix arch files.
But I haven't reviewed the rest.
[Stating the obvious: This PR causes a pile of failures in CI]

Vudentz commented on arch/arc/core/mpu/arc_core_mpu.c:
```diff
@@ -47,7 +47,7 @@ void configure_mpu_thread(struct k_thread *thread)
 {
 #if defined(CONFIG_USERSPACE)
-	if (thread->thread_base.user_options & K_USER) {
+	if ((thread->thread_base.user_options & K_USER) != 0) {
```
I wonder if wouldn't be better to have a macro to check for a mask, something like:

#define check_mask(_value, _mask) \
             (_value & _mask) != 0

ceolin replied:
we can have something for this indeed.

Vudentz commented on arch/nios2/core/fatal.c:
```diff
@@ -187,7 +187,7 @@ FUNC_NORETURN void _Fault(const NANO_ESF *esf)
 	printk("reason: %s\n", cause_str(cause));
 #endif
-	if (BIT(cause) & NIOS2_BADADDR_CAUSE_MASK) {
+	if ((BIT(cause) & NIOS2_BADADDR_CAUSE_MASK) != 0) {
```
It seems we could just use BIT(cause) == NIOS2_BADADDR_CAUSE_MASK here.

andrewboie replied:
no, NIOS2_BADADDR_CAUSE_MASK is more than one bit

Vudentz commented on arch/posix/core/posix_core.c:
```diff
@@ -131,7 +131,7 @@ static void posix_wait_until_allowed(int this_th_nbr)
 		pthread_cond_wait(&cond_threads, &mtx_threads);
 
-		if (threads_table &&
+		if ((threads_table != NULL) &&
```
Should this be the other way around like: NULL != pointer, that way we avoid the risk of assignments instead of comparisons? (Though I must say I found this very silly rule since most people do understand these expression nowadays, and I don't think we support any architecture which NULL is not 0. )

andrewboie replied:
I don't think MISRA-C requires Yoda conditions

ceolin replied:
The guideline does not give a big rationale for this rule. It just says that strong typing requires controlling expression of a if statement have a boolean type.

himanshujha199640 reviewed (commented):
Take a look at this:
https://gist.github.com/himanshujha199640/f2b05c8da5f83f604b48d733e18696a0

Your current implementation is not well structured.

For more help on bools: https://github.com/torvalds/linux/tree/master/scripts/coccinelle/misc
search for `bool` scripts and you need to do the reverse with those scripts since in linux
we prefer `if(!foo)` than `if(foo == NULL)`

himanshujha199640 commented:
FYI @JuliaLawall

ceolin commented on include/misc/rb.h:
```diff
@@ -180,8 +180,8 @@ struct rbnode *_rb_foreach_next(struct rbtree *tree, struct _rb_foreach *f);
-	     (node = CONTAINER_OF(_rb_foreach_next(tree, &__f),		\
-				  __typeof__(*(node)), field));		\
+	     ({node = CONTAINER_OF(_rb_foreach_next(tree, &__f),	\
+			   __typeof__(*(node)), field); node != NULL; }); \
```
Agree! and this change even violates other advisory rule, though I didn't see other way to do this in a macro. We can change it to be a function and receive a callback, but this will trigger bigger changes.
I'm open for suggestions here since we have more similar cases.

ceolin commented:
@himanshujha199640 Yeah, that's why I had asked your help in the previous commit. I know that this can be much better, but I always have problem doing more elaborate scripts then I end up being really redundant :)

himanshujha199640 commented:
> @himanshujha199640 Yeah, that's why I had asked your help in the previous commit. I know that this can be much better, but I always have problem doing more elaborate scripts then I end up being really redundant :)

No worries, even I end up with redundancies until Julia points me the correct way.
I tried a `context` only script:
https://gist.github.com/himanshujha199640/9b40392a1f5d082013c1d5af3d6075ec

You may want to take a look at various isomorphim that I have disabled for effective matching:
https://github.com/coccinelle/coccinelle/blob/175de16bc7e535b6a89a62b81a673b0d0cd7075c/standard.iso

Use disjuctions and dots patterns to compact your script and increase the confidence. At least try to
make the confidence high and remove false positives. If there are any weird false positvies then let
me know.

himanshujha199640 commented on scripts/coccinelle/boolean-expression.cocci:
```diff
@@ -0,0 +1,357 @@
+typedef u16_t;
+typedef uint32_t;
+typedef u32_t;
+{uint8_t,u8_t,uint16_t,u16_t,uint32_t,u32_t,unsigned int,int,size_t,unsigned} I;
```
Why unsigned only ?

himanshujha199640 commented on scripts/coccinelle/boolean-expression.cocci:
```diff
@@ -0,0 +1,357 @@
+@@
+
+- if (X && E)
++ if ((X != NULL) && E)
```
Use `...` instead as your case won't catch:
```
int main (void)
{
	unsigned int *foo; 
	if (foo && baz && boat)
		bar++;
}
```

ceolin replied:
use `...` instead of ? the expression `E` ?

ceolin replied:
yeah, there are several cases, that is why I put low confidence. Also I got some problems of comparing non-pointers with NULL. Last but no least, some problems with adding `{` in cases like:
```
else if (....)  {
}
```
becoming
```
else { if (...) {
         }
}
```

himanshujha199640 commented on scripts/coccinelle/boolean-expression.cocci:
```diff
@@ -0,0 +1,357 @@
+constant C;
+@@
+
+- if (I & C)
```
Any particular reson for using `identifier` and `constant`.
I assume your main intention is to change
```
- if (A & B)
+ if ((A & B) != 0)
```
without caring about the type of A & B.
So, I think both A, B can be simply `expression` metavariables.

himanshujha199640 reviewed (commented):
One can also think about cases such as:
```
int main (void)
{
    if (foo->bar && temp)
         baz++;
}
```
And numerous other cases as well. :)

Now I imagine the complexity of the script.

JuliaLawall commented:
On Wed, 28 Nov 2018, Himanshu Jha wrote:

>
> @himanshujha199640 commented on this pull request.
>
> One can also think about cases such as:
>
> int main (void)
> {
>     if (foo->bar && temp)
>          baz++;
> }
>
> And numerous other cases as well. :)
>
> Now I imagine the complexity of the script.
>
> ____________________________________________________________________________
>
> In scripts/coccinelle/boolean-expression.cocci:
>
> > +statement S;
> +@@
> +
> +- if (X) {
> ++ if (X != NULL) {
> +S
> +}
> +
> +@depends on !(file in "ext")@
> +typedef uint8_t;
> +typedef u8_t;
> +typedef uint16_t;
> +typedef u16_t;
> +typedef uint32_t;
> +typedef u32_t;
> +{uint8_t,u8_t,uint16_t,u16_t,uint32_t,u32_t,unsigned int,int,size_t,unsigne
> d} I;
>
> Why unsigned only ?
>
> ____________________________________________________________________________
>
> In scripts/coccinelle/boolean-expression.cocci:
>
> > +expression *E;
> +@@
> +
> +- 0
> ++ NULL
> +  != E
> +
> +
> +@depends on !(file in "ext")@
> +idexpression *X;
> +expression E;
> +statement S;
> +@@
> +
> +- if (X && E)
> ++ if ((X != NULL) && E)
>
> Use ... instead as your case won't catch:

There are some problems with the ... idea.  I am looking into it.

julia

>
> int main (void)
> {
> 	unsigned int *foo;
> 	if (foo && baz && boat)
> 		bar++;
> }
>
> ____________________________________________________________________________
>
> In scripts/coccinelle/boolean-expression.cocci:
>
> > +statement S;
> +expression E;
> +constant C;
> +@@
> +
> +- if (C & E)
> ++ if ((C & E) != 0)
> +  S
> +
> +@depends on !(file in "ext")@
> +statement S;
> +identifier I;
> +constant C;
> +@@
> +
> +- if (I & C)
>
> Any particular reson for using identifier and constant.
> I assume your main intention is to change
>
> - if (A & B)
> + if ((A & B) != 0)
>
> without caring about the type of A & B.
> So, I think both A, B can be simply expression metavariables.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the
> thread.[AAesmjN8tHOW5Ays-tKDIRAj9PE8gi1sks5uzs6EgaJpZM4Y0JpT.gif]
>
>
>

JuliaLawall commented:
On Wed, 28 Nov 2018, Flávio Vinicius Alvares Ceolin wrote:

>
> @ceolin commented on this pull request.
>
> ____________________________________________________________________________
>
> In scripts/coccinelle/boolean-expression.cocci:
>
> > +expression *E;
> +@@
> +
> +- 0
> ++ NULL
> +  != E
> +
> +
> +@depends on !(file in "ext")@
> +idexpression *X;
> +expression E;
> +statement S;
> +@@
> +
> +- if (X && E)
> ++ if ((X != NULL) && E)
>
> use ... instead of ? the expression E ?

It should be possible to write

- X
+ (X != NULL)
  && ...

to have the transformation apply anywhere in a top-level string of &&,
including no &&.  At the moment, though, it is not working.  I need to fix
this.

julia



>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the
> thread.[AAesmrgUgpouzXrvDrtNw_gQey8M8a_tks5uzxTUgaJpZM4Y0JpT.gif]
>
>
>

himanshujha199640 commented on lib/libc/newlib/libc-hooks.c:
```diff
@@ -189,7 +189,7 @@ FUNC_ALIAS(_fstat, fstat, int);
 {
 	_write(1, "exit\n", 5);
-	while (1) {
+	while (true) {
```
Minor Nit:
I couldn't resist myself on this change. But I would like to point that
`true` is defined in `stddbool.h` which **ultimately** expands to
integer constant 1.
http://pubs.opengroup.org/onlinepubs/009604499/basedefs/stdbool.h.html

With this change I saw a shippable build failure because now we need to
explicitly add the header file wherever required.

Commit:
arc: cache: Change dcache_available signature

This function returns an essential boolean type, just changing the
signature to return a boolean type.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
kernel: userspace: Change _thread_idx_alloc to return bool

Make this function return an essential boolean type.

MISRA-C rule 14.4

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
kernel: Make statements evaluate boolean expressions

MISRA-C requires that the if statement has essentially Boolean type.

MISRA-C rule 14.4

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
kernel: Change is_condition_met signature

Change this function to return a boolean type.

MISRA-C rule 14.4

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
arch: x86: Make statements evaluate boolean expressions

MISRA-C requires that the if statement has essentially Boolean type.

MISRA-C rule 14.4

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
log: Make statements evaluate boolean expressions

MISRA-C requires that the if statement has essentially Boolean type.

MISRA-C rule 14.4

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
arch: arm: Make statements evaluate boolean expressions

MISRA-C requires that the if statement has essentially Boolean type.

MISRA-C rule 14.4

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
arch: arc: Make statements evaluate boolean expressions

MISRA-C requires that the if statement has essentially Boolean type.

MISRA-C rule 14.4

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
kernel: Change _is_thread_essential to return bool

Change this function to return a boolean type.

MISRA-C rule 14.4

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
kernel: Change k_is_in_isr to return bool

Change this function to return a boolean type.

MISRA-C rule 14.4

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
lib: posix: Make statements evaluate boolean expressions

MISRA-C requires that the if statement has essentially Boolean type.

MISRA-C rule 14.4

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
soc: x86: Make statements evaluate boolean expressions

MISRA-C requires that the if statement has essentially Boolean type.

MISRA-C rule 14.4

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
soc: arm: Make statements evaluate boolean expressions

MISRA-C requires that the if statement has essentially Boolean type.

MISRA-C rule 14.4

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

ceolin commented:
other thing, I remove some subsystems from this change too.

ceolin commented:
@andrewboie can you please take a look again in this pr ?

nordic-krch reviewed (approved):
+1 for log part
